### DISCUSSION TITLE
AsyncSession poor performance

### DISCUSSION THREAD

**Discussion by jvsnap** [sqlalchemy/sqlalchemy#8947] [Usage Questions]
I got a question about how to properly use create_async_engine. The following code takes about 1.5 sec to query into my production database, it is a simple geospatial query.

```python
import time
import asyncio

from asyncio import run as aiorun

from sqlalchemy import func, select
from sqlalchemy.orm import sessionmaker
from sqlalchemy import create_engine

from typing import Any, Dict, List

import models


def get_session():
    engine = create_engine(
        "postgresql+psycopg2://postgres:postgres@127.0.0.1:5431/db",
        pool_size=20,
        max_overflow=0,
        pool_pre_ping=True,
    )
    return sessionmaker(bind=engine)

sess = get_session()
session = sess()

_geom = "SRID=4269;POLYGON(( ... ))"

def main():
    stmt = select(
        models.FloodHazardAreas, func.ST_Intersection(
            models.FloodHazardAreas.geom, _geom)
    ).where(func.ST_Intersects(models.FloodHazardAreas.geom, _geom))

    rows = session.execute(stmt).scalars()

main()
```

But when I try to adapt the same logic by using AsyncSession then it takes more than 320 sec to complete the same query as above.

```python
import time
import asyncio

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker

import models


engine = create_async_engine(
    "postgresql+asyncpg://postgres:postgres@127.0.0.1:5431/db",
    echo=True,
)

async_session = sessionmaker(
    engine, class_=AsyncSession, expire_on_commit=False)

_geom = "SRID=4269;POLYGON(( ... ))"

async def async_main():    
    async with async_session() as conn:
        stmt = select(
            models.GeoTable, func.ST_Intersection(
                models.GeoTable.geom, _geom)
        ).where(func.ST_Intersects(models.GeoTable.geom, _geom))

        results = await conn.execute(stmt)

asyncio.run(async_main())
```

Thanks!

**Comment by CaselIT** [comment]
Hi,

Since the examples are not runnable, logging would help.
Is seems strange that this is an sqlalchemy issue, since there doesn't seem to be anything in particular going on.

Not sure if maybe the cause may be the reflection logic of asyncpg ? https://github.com/MagicStack/asyncpg/blob/master/asyncpg/introspection.py

how much time does it take running the same query with asyncpg directly?

**Reply by jvsnap** [reply]
Thanks @CaselIT when using `asyncpg` it takes about 1.2 secs

**Reply by zzzeek** [reply]
we use prepared statements with asyncpg.  that might be pulling in their introspection query

**Reply by zzzeek** [reply]
the best way to know exactly the problem is to [use the Python profiler](https://docs.sqlalchemy.org/en/14/faq/performance.html#code-profiling)

**Reply by CaselIT** [reply]
also it may be worth trying to use sqlalchemy 2 beta with psycopg that has an async version

**Reply by zzzeek** [reply]
psycopg has performance problems of its own :)    curious to see the profile result on this one

**Reply by CaselIT** [reply]
Sure, but not a 200x performance problem wrt psycopg2 :)

**Reply by zzzeek** [reply]
im assuming something is going wrong in the asycnpg case.  if there is in fact some thing we're doing w/ prepared statement that is doing that, or something geoalchemy does, no idea.
